File: perceptron.js
var _ = require('lodash'),
  fs = require('fs'),
  stdin = process.stdin,
  PT = require('./tools.js');

function training (weights, domain, size, iterations, lRate) {
  for (var i=0; i<iterations; i++) {
    var totalErr = 0.0;
    var curv = (1 - i/iterations);
    var adjLRate = lRate * curv * curv;
    
    _.each(domain, function (pattern, j) {
      var input = _.map(PT.initialArray(size), function (x, k) {
        return pattern[k];
      });
      var expected = _.last(pattern);
      
      weights = _.map(weights, function (w, l) {
        var exp = -1.0;
        if (expected == l) exp = 1.0;
        
        var output = PT.getOutput(w, input);
        var err = exp - output;
        totalErr += Math.abs(err);
        
        return PT.updateWeights(size, w, input, err, adjLRate);
      });
    });
    
    console.log(adjLRate);
    console.warn('> epoch=%d, error=%d', i, totalErr)
  }
  return weights;
}

function testing (weights, domain, size) {
  var correct = 0;
  
  _.each(domain, function (pattern, i) {
    var input = _.map(PT.initialArray(size), function (x, j) {
      return pattern[j];
    });
    
    var best = 0;
    var select = 0;
    
    _.each(weights, function (w, j) {
      var activation = PT.activate(w, input);
      if (activation >= best) {
        best = activation;
        select = j;
      }
    });
    
    var answer = _.last(pattern);
    console.log('ai: %d, answer: %d', select, answer);
    
    if (select == answer) correct += 1;
  });
  
  return correct;
}

var classSize = 2,
  inputSize = 2,
  iterations = 3000,
  learningRate = 0.1,
  weightList = [];
  
for (var i=0; i<classSize; i++) {
  weightList.push(PT.initializeWeights(inputSize));
}
  
function menu () {
  console.log('[L]earning. [T]esting. [Q]uit.');
  
  PT.ask('> ', /.+/, function (data) {
    if ( data.toLowerCase() == 'q' ) {
      process.exit();
    }
    else if ( data.toLowerCase() == 'l' ) {
      var domain = [];
      fs.readFile('./od.tra', 'utf8', function(err, data) {
        if (err) {
          console.log('File not found.');
          return;
        }
        var lines = data.split('\n');
        
        _.each(lines, function (line, i) {
          var words = line.split(',');
          domain.push(words);
        });
        
        weightList = training(weightList, domain, inputSize, iterations, learningRate);
        menu();
      });
    }
    else if ( data.toLowerCase() == 't' ) {
      var domain = [];
      fs.readFile('./od.tes', 'utf8', function(err, data) {
        if (err) {
          console.log('File not found.');
          return;
        }
        var lines = data.split('\n');
        
        _.each(lines, function (line, i) {
          var words = line.split(',');
          domain.push(words);
        });
        
        var c = testing(weightList, domain, inputSize);
        
        console.log('result : %d / %d', c, domain.length);
        menu();
      });
    }
    else {
      menu();
    }
  });
}

menu();

File: perceptron.py
import random

def random_vector(minmax):
	arr = [None]*len(minmax)
	for i in range(0, len(minmax)):
		arr[i] = minmax[i][0] + ((minmax[i][1] - minmax[i][0]) * random.random())
	return arr
	
def initialize_weights(problem_size):
	minmax = [[-1.0,1.0]]*(problem_size + 1)
	return random_vector(minmax)
	
def update_weights(num_inputs, weights, input, out_exp, out_act, l_rate):
	for i in range(0,num_inputs):
		weights[i] += l_rate * (out_exp - out_act) * input[i]
	weights[num_inputs] += l_rate * (out_exp - out_act) * 1.0

def activate(weights, vector):
	sum = weights[len(weights)-1] * 1.0
	for i in range(0, len(vector)):
		w = weights[i]
		v = vector[i]
		sum += w * v
	return sum

def transfer(activation):
	if activation >= 0:
		return 1.0
	return 0.0

def get_output(weights, vector):
	activation = activate(weights, vector)
	return transfer(activation)

def train_weights(weights, domain, num_inputs, iterations, lrate):
	for i in range(0, iterations):
		error = 0.0
		for pattern in domain:
			input = [None]*num_inputs
			for j in range(0,len(input)):
				input[j] = float(pattern[j])
			expected = int(pattern[len(pattern)-1])
			
			for k in range(0,len(weights)):
				e = 0
				if expected == k:
					e = 1
				w = weights[k]
				output = get_output(w, input)
				error += abs(output - e)
				update_weights(num_inputs, w, input, e, output, lrate)
		print("> epoch=%d, error=%f" % (i, error))
# 30: 1546
# 500: 1542

def test_weights(weights, domain, num_inputs):
	correct = 0
	for pattern in domain:
		input_vector = [None]*num_inputs
		for i in range(0,len(input_vector)):
			input_vector[i] = float(pattern[i])
		
		best = 0
		select = 0
		for i in range(0,len(weights)):
			activation = activate(weights[i], input_vector)
			if activation >= best:
				best = activation
				select = i
		answer = int(pattern[len(pattern)-1])
		print('ai: %d, anwer: %d' % (select, answer))
		if select == answer:
			correct += 1
	return correct

inputs = 64
iterations = 5
learning_rate = 0.1
weights = []

while 1:
	print('1 for Learning, 2 for Testing, quit for another')
	mode = input('>> ');

	if mode == 1:
		for i in range(0,10):
			weights.append(initialize_weights(inputs))
			
		print('Learning')
		print('input file name')
		fileName = input('>> ')
		f = open(fileName, 'r')

		if not f:
			print('file not found')
			continue;

		domain = []
		while 1:
			line = f.readline()
			if not line:
				break			
			words = line.split(',')
			domain.append(words)
# 			classifier = words[len(words)-1].replace('\n', '')
# 			print(classifier)
		train_weights(weights, domain, inputs, iterations, learning_rate)

		f.close()
	elif mode == 2:
		print('Testing')
		print('input file name')
		fileName = input('>> ')
		f = open(fileName, 'r')

		if not f:
			print('file not found')
			continue;

		domain = []
		while 1:
			line = f.readline()
			if not line:
				break			
			words = line.split(',')
			domain.append(words)
# 			classifier = words[len(words)-1].replace('\n', '')
# 			print(classifier)
		c = test_weights(weights, domain, inputs)
		
		print("result: %d / %d" % (c, len(domain)))

		f.close()
	else:
		print('quit')
		exit()
